Fix TFile gap coalescing

jblomer opened the pull request:
Fixes #19245

jblomer committed:
[io] fix TFile gap coalescing

Prevent merged free segments from becoming bigger than 2GB. Requires
additional code to ensure that gaps at the end of the file are always
merged.

silverweed commented on io/io/src/TFile.cxx:
```diff
@@ -1474,27 +1474,55 @@ Bool_t TFile::IsOpen() const
+
+   Long64_t nbytesl = nlast - nfirst + 1;
+   if (nbytesl > TFree::kMaxGapSize)
+      nbytesl = TFree::kMaxGapSize;
```
If I understand correctly, here we are "sacrificing" `nbytesl - kMaxGapSize` bytes and leaving them unindexed; shouldn't we also add a new free segment right after `tail` to encompass them if this is not at the end of the file?

pcanal replied:
(side note, this lost was already present is the current code).

silverweed commented on io/io/src/TFree.cxx:
```diff
@@ -70,17 +70,23 @@ TFree *TFree::AddFree(TList *lfree, Long64_t first, Long64_t last)
    while (idcur) {
```
This should be a `do / while` but I guess it's fine either way...

pcanal commented on io/io/inc/TFree.h:
```diff
@@ -23,14 +23,16 @@
 
 public:
+   /// Prevent gap coalescing from exceeding 2 GB, since the gap size is stored as a signed 4 bytes integer.
+   /// The true limit is 2 GiB but for historical reasons we stick to the 2GB limit.
```
Should we add a hint of where the historical reasons can be found? :)

pcanal replied:
```suggestion
   /// The true limit is 2 GiB but for historical reasons (previous implement of TFree) we stick to the 2GB limit.
```

pcanal commented on io/io/src/TFree.cxx:
```diff
@@ -70,17 +70,23 @@ TFree *TFree::AddFree(TList *lfree, Long64_t first, Long64_t last)
+         if (idnext == 0)
+            return idcur;
+
+         // Is the next segment adjacent to the newly merged one (and not too big)
```
Either
```suggestion
         //  Continue only if the next segment adjacent to the newly merged one (and not too big)
```
or 
Either
```suggestion
         //  Return unless the next segment adjacent to the newly merged one (and not too big)
```

jblomer commented:
... still some non-understood issues on recovery, working on that.

jblomer committed:
[io] always recreate free list in TFile::Recover()

This avoids tainting the recovered free list by stale entries that were
read before. It also allows for stronger assertions in TFree::AddFree().

jblomer commented:
Replaced by https://github.com/root-project/root/pull/22481
